[iOS 17] [OOP Keyboard] Scrolling to reveal the selection is broken in Mail compose

whsieh opened the pull request:
#### eb551731bdda4d96b22724af1703ba169df61f0b
<pre>
[iOS 17] [OOP Keyboard] Scrolling to reveal the selection is broken in Mail compose
<a href="https://bugs.webkit.org/show_bug.cgi?id=258573">https://bugs.webkit.org/show_bug.cgi?id=258573</a>
rdar://110139313

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

We currently call `-[WKContentView reloadInputViews]` twice whenever we focus an editable element on
iOS. While this *should* be benign, it causes UIKit to dispatch redundant sets of `KeyboardWillShow`
and `KeyboardDidShow` notifications before the keyboard animation has even started, when out-of-
process keyboard is enabled.

Importantly, this redundant set of notifications causes us to not scroll to reveal the selection
when tapping into Mail compose to bring up the keyboard, since the keyboard height is still 0 (i.e.
the keyboard is not covering anything) when the earliest set of keyboard notifications arrives.

While the redundant notification dispatching likely still needs to be fixed in UIKit, it seems
wasteful to unnecessarily trigger `-reloadInputViews` in WebKit, so we can address the issue for now
by fixing this.

* LayoutTests/editing/selection/ios/scroll-to-reveal-selection-in-mail-compose-expected.txt: Added.
* LayoutTests/editing/selection/ios/scroll-to-reveal-selection-in-mail-compose.html: Added.

Augment an existing layout test that was originally intended to simulate Mail compose, such that it
more closely follows Mail compose with the software keyboard — i.e.:

- The web view is editable.
- First responder keyboard avoidance scrolling is disabled.
- The input accessory view is suppressed.

* LayoutTests/editing/selection/ios/scroll-to-reveal-selection-with-keyboard-avoidance-disabled-expected.txt: Removed.
* LayoutTests/editing/selection/ios/scroll-to-reveal-selection-with-keyboard-avoidance-disabled.html: Removed.
* LayoutTests/editing/selection/ios/update-selection-after-iframe-scroll-expected.txt:
* LayoutTests/editing/selection/ios/update-selection-after-iframe-scroll.html:

Drive-by-fix a flaky layout test, which fails if run after some other layout tests (discovered while
debugging nearby test failures), by (1) ensuring that the test begins after the `iframe` finishes
loading, and (2) the selection in editable content is always set properly, by setting it after
starting the input session — the initial tap sometimes causes the selection to collapse right after
the click handler selects a range.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _showKeyboard]):

Also remove an unnecessary check for `PLATFORM(WATCHOS)`, now that this method is only invoked from
non-watchOS codepaths.

(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):

See comments above.

* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:

Add a new test option to suppress the input accessory view when starting an input session; necessary
to satisfy the conditions required to trigger this bug.

(WTR::TestOptions::suppressInputAccessoryView const):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView _webView:willStartInputSession:]):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::platformResetStateToConsistentValues):

Canonical link: <a href="https://commits.webkit.org/265556@main">https://commits.webkit.org/265556@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f22a52d930b0ca288becdbe9fe9a29a895ee0b4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11227 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11437 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11769 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12878 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10707 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11248 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13816 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11422 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13621 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11388 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12272 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/9488 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/13296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9566 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10164 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17362 "layout-tests (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10636 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10319 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13538 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10748 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8835 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9918 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10055 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2687 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14192 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10601 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
```diff
@@ -545,6 +546,7 @@ - (void)_webView:(WKWebView *)webView willStartInputSession:(id <_WKFormInputSes
 {
     if (self.willStartInputSessionCallback)
         self.willStartInputSessionCallback();
+    inputSession.accessoryViewShouldNotShow = self.suppressInputAccessoryView;
```
We want this to be `true` right? Are we setting that anywhere?

whsieh replied:
Yep! The layout test specifies my new test option:

```
<!DOCTYPE html> <!-- webkit-test-runner [ useFlexibleViewport=true suppressInputAccessoryView=true ] -->
…
```

aprotyas reviewed (approved):
One question but looks good 👍

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -6926,9 +6926,7 @@ - (void)_showKeyboard
     if (self.isFirstResponder && !_suppressSelectionAssistantReasons)
         [_textInteractionAssistant activateSelection];
 
-#if !PLATFORM(WATCHOS)
     [self reloadInputViews];
-#endif
```
Intentional change?

whsieh replied:
Ah, yes — (it wasn't super clear from the commit message, but I adjusted this since we no longer call `-_showKeyboard` on watchOS). I can add a sentence to the commit message to make this clearer.

whsieh committed:
[iOS 17] [OOP Keyboard] Scrolling to reveal the selection is broken in Mail compose
https://bugs.webkit.org/show_bug.cgi?id=258573
rdar://110139313

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

We currently call `-[WKContentView reloadInputViews]` twice whenever we focus an editable element on
iOS. While this *should* be benign, it causes UIKit to dispatch redundant sets of `KeyboardWillShow`
and `KeyboardDidShow` notifications before the keyboard animation has even started, when out-of-
process keyboard is enabled.

Importantly, this redundant set of notifications causes us to not scroll to reveal the selection
when tapping into Mail compose to bring up the keyboard, since the keyboard height is still 0 (i.e.
the keyboard is not covering anything) when the earliest set of keyboard notifications arrives.

While the redundant notification dispatching likely still needs to be fixed in UIKit, it seems
wasteful to unnecessarily trigger `-reloadInputViews` in WebKit, so we can address the issue for now
by fixing this.

* LayoutTests/editing/selection/ios/scroll-to-reveal-selection-in-mail-compose-expected.txt: Added.
* LayoutTests/editing/selection/ios/scroll-to-reveal-selection-in-mail-compose.html: Added.

Augment an existing layout test that was originally intended to simulate Mail compose, such that it
more closely follows Mail compose with the software keyboard — i.e.:

- The web view is editable.
- First responder keyboard avoidance scrolling is disabled.
- The input accessory view is suppressed.

* LayoutTests/editing/selection/ios/scroll-to-reveal-selection-with-keyboard-avoidance-disabled-expected.txt: Removed.
* LayoutTests/editing/selection/ios/scroll-to-reveal-selection-with-keyboard-avoidance-disabled.html: Removed.
* LayoutTests/editing/selection/ios/update-selection-after-iframe-scroll-expected.txt:
* LayoutTests/editing/selection/ios/update-selection-after-iframe-scroll.html:

Drive-by-fix a flaky layout test, which fails if run after some other layout tests (discovered while
debugging nearby test failures), by (1) ensuring that the test begins after the `iframe` finishes
loading, and (2) the selection in editable content is always set properly, by setting it after
starting the input session — the initial tap sometimes causes the selection to collapse right after
the click handler selects a range.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _showKeyboard]):

Also remove an unnecessary check for `PLATFORM(WATCHOS)`, now that this method is only invoked from
non-watchOS codepaths.

(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):

See comments above.

* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:

Add a new test option to suppress the input accessory view when starting an input session; necessary
to satisfy the conditions required to trigger this bug.

(WTR::TestOptions::suppressInputAccessoryView const):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView _webView:willStartInputSession:]):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::platformResetStateToConsistentValues):

Canonical link: https://commits.webkit.org/265556@main
